### PR TITLE
Announce quickfix only if there were edits

### LINKED ISSUE #12940: With `-quickfix` enabled, `[rewritten by quickfix]` gets printed for every error, regardless of whether or not it was actually fixed/fixable
## Reproduction steps

Scala version: 2.13.12

in Test.scala:
```scala
import foo.bar
```
Compiling with `-quickfix:any` prints
```
Test.scala:1: error: [rewritten by -quickfix] not found: object foo
import foo.bar
       ^
1 error
```
## Problem
It says the "not found: object foo" error was "rewritten by -quickfix" but it wasn't, it's not a `quickfixable` error

### LINKED ISSUE #12941: quickfixes sometimes don't get successfully applied
## Reproduction steps

Scala version: 2.13.12

in Test1.scala:
```scala
trait Trait[A] {
  def foo: Option[A]
}
object Test1 extends Trait[Any] {
  def foo = None
}
```
in Test2.scala:
```scala
object Test2 {
  def one = Test1.foo
}
```

`scala-2.13.12/bin/scalac -quickfix:any -Xsource:3 Test2.scala Test1.scala` prints
```
Test1.scala:5: error: [rewritten by -quickfix] under -Xsource:3, inferred Option[Any] instead of None.type
Scala 3 migration messages are errors under -Xsource:3. Use -Wconf / @nowarn to filter them or add -Xmigration to demote them to warnings.
Applicable -Wconf / @nowarn filters for this fatal warning: msg=<part of the message>, cat=scala3-migration, site=Test1.foo
  def foo = None
      ^
1 error
```
_but it doesn't actually change Test1.scala at all_. If you swap the order the files are provided in the command line, or if you remove the `.foo` in Test2.scala, the quickfix for Test1.scala gets applied as expected. I.e., `scala-2.13.12/bin/scalac -quickfix:any -Xsource:3 Test1.scala Test2.scala` has the same output but it actually updates Test1.scala.

## Problem
The order of source files passed to scalac shouldn't affect whether quickfixes are applied or not

### PR DESCRIPTION
**Pull request by som-snytt**
Fixes scala/bug#12940
Fixes scala/bug#12941

Needs a test.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/Reporting.scala [hunk 1]
```diff
@@ -93,12 +93,12 @@ trait Reporting extends internal.Reporting { self: ast.Positions with Compilatio
       else false
 
     private def registerErrorTextEdit(pos: Position, msg: String, actions: List[CodeAction]): Boolean = {
-      val matches = quickfixFilters.exists({
-        case MessageFilter.Any => true
+      val matches = quickfixFilters.exists {
+        case MessageFilter.Any => actions.exists(_.edits.nonEmpty)
         case mp: MessageFilter.MessagePattern => mp.check(msg)
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/Reporting.scala:99]: Also for other filters:

```
➜ sandbox sc Test.scala '-quickfix:msg=type mismatch'
Test.scala:2: error: [rewritten by -quickfix] type mismatch;
 found   : Int(1)
 required: String
  def f: String = 1
                  ^
1 error
```

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/Reporting.scala:99]: Thanks, I'll follow up in our morning / post-sunrise.

Comment (som-snytt) [reply src/compiler/scala/tools/nsc/Reporting.scala:99]: Or I better do it now before I fall asleep and forget everything in dreams.

**Review by lrytz — APPROVED** [review]
👍 thanks!

File: src/compiler/scala/tools/nsc/typechecker/Namers.scala [hunk 2]
```diff
@@ -1122,7 +1122,7 @@ trait Namers extends MethodSynthesis {
           val leg = legacy.toString
           val help = if (pts != leg) s" instead of $leg" else ""
           val msg = s"under -Xsource:3-cross, the inferred type changes to $pts$help"
-          val src = currentUnit.source
+          val src = tree.pos.source
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Namers.scala:1125]: Ah I didn't realize this PR also fixes https://github.com/scala/bug/issues/12941, I guess that fix is here? I'll add a test.

**Pull request merged**